lib/rules/prefer-destructuring.js

```diff
@@ -0,0 +1,128 @@
+/**
+ * @fileoverview Prefer destructuring from arrays and objects
+ * @author Alex LaFroscia
+ */
+"use strict";
+
+//------------------------------------------------------------------------------
+// Rule Definition
+//------------------------------------------------------------------------------
+
+module.exports = {
+    meta: {
+        docs: {
+            description: "require destructuring from arrays and/or objects",
+            category: "ECMAScript 6",
+            recommended: false
+        },
+
+        schema: [
+            {
+                type: "object",
+                properties: {
+                    array: {
+                        type: "boolean"
+                    },
+                    object: {
+                        type: "boolean"
+                    }
+                }
+            }
+        ]
+    },
+    create(context) {
+
+        let checkArrays = true;
+        let checkObjects = true;
+        const options = context.options[0];
+
+        if (options) {
+            if (options.hasOwnProperty("array")) {
+                checkArrays = options.array;
+            }
+
+            if (options.hasOwnProperty("object")) {
+                checkObjects = options.object;
+            }
+        }
+
+        //--------------------------------------------------------------------------
+        // Helpers
+        //--------------------------------------------------------------------------
+
+        /**
+         * Determines if the given node node is accessing an array index
+         *
+         * This is used to differentiate array index access from object property
+         * access.
+         *
+         * @param {ASTNode} node the node to evaluate
+         * @returns {boolean} whether or not the node is an integer
+         */
+        function isArrayIndexAccess(node) {
+            return Number.isInteger(node.property.value);
```

Comment:
This check looks fine and it's correct behavior, but as far as usability is concerned, I'm thinking it might be annoying to get a warning for code like this:

```js
const foo = array[100];
```

The correct fix for that code would be:

```js
const [ , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , , foo] = array;
```